Add fwupd_device_get_id_parent()

hughsie opened the pull request:
This allows us to find out the logical parent device, for instance in composite
devices with more than one firmware image for a single device.

superm1 commented on src/fu-device.c:
```diff
@@ -164,6 +165,50 @@ fu_device_set_alternate (FuDevice *device, FuDevice *alternate)
+ **/
+void
+fu_device_set_parent (FuDevice *device, FuDevice *parent)
+{
```
I think I would rather see these send and return parent ID's and let the daemon look up corresponding FuDevice. My logic is that it will be easier to predict and associate cross plugin which device is your parent if you know how the device is logically built or connected in advance.

hughsie replied:
Yes and no. I do think having the FuDevice `GObject` of the parent is a useful thing to plugins, but I agree the daemon could do more matching up devices. I was intending to use `fu_device_set_parent()` directly in plugins where all the logical devices are in the "same" plugin (e.g. `unifying`) and for where the logical devices are spread around between plugins I'd add something like `fu_device_set_parent_with_guid(d,guid_string)` and/or `fu_device_set_parent_with_metdata(d,key,value)` and /then/ let the daemon call `fu_device_set_parent` so the actual object is usable by the "child" plugin when and if the parent shows up.

superm1 replied:
Okay that works fine to me then.

hughsie replied:
I was thinking more about the "find me a parent" case. I guess the most useful thing would be `fu_device_set_parent_with_guid` where the GUID is stored in the child `FuDevice` in an array (in the case the parent might be different things). Then we have two cases:

* parent added, then child
* child added, then parent

In both cases we just need to rescan the "adoption" list when a device is added. I think it's fair to say we won't allow re-parenting or having more than one parent. Would this work for your dock case?

superm1 replied:
Yeah I think that would work.

hughsie replied:
Something like the patch below?

superm1 commented on src/fu-device-list.c:
```diff
@@ -451,6 +491,9 @@ fu_device_list_add (FuDeviceList *self, FuDevice *device)
 	}
 
+	/* adopt any required children, which may or may not already exist */
+	fu_device_list_adopt_children (self, device);
```
I think it would be better to do this a little earlier (before all plugins are notified about the new device).  This would mean that plugins will be able to know about devices coming in with a parent attached.

hughsie replied:
So move the adoption into FuEngine rather than the DeviceList? I guess that makes sense.

hughsie replied:
Fixed.

superm1 commented:
Just my one comment above, otherwise looks and works fine.
I added a patch to thunderbolt that uses this to logically organize TBT under the Dell TB16 like how it would work.
(CC @YehezkelShB if you would like to look)

superm1 commented:
> I was thinking a bit earlier about how we're going to present this in the CLI -- do you think it makes sense to create a dummy USB (or firewire, or PCI, for that matter) controller for instance, so that the various USB devices are listed under that? I don't think a "local computer" makes sense as everything without a parent would just be considered a child of the non-existent root node.

Yes I do think dummy top level bus parents make sense.  I'm fine with it being presented under a separate command too (eg `get-topology`) if you don't want to break the existing output style from get-devices

hughsie commented:
I faked the parent relationships to test how something like get-topology would look:

hughsie commented:
![screenshot from 2018-05-01 21-49-07](https://user-images.githubusercontent.com/151380/39493089-82fdce86-4d89-11e8-8a0a-c977ed07defd.png)

superm1 commented:
Looks pretty good to me.
I think in the `get-topology` the one other thing you would want to show is the hash in parentheses as that's what is used in all the other commands for installing to something, getting details about it etc.

YehezkelShB commented:
@superm1
Makes sense.
On new devices (TR), there is a field in DROM that should be used to mark multiple controllers inside the same device, so we'll be able to use this too instead of hard-coding IDs for each such device.
(Before this DROM field, there was a byte somewhere else in the FW that should have been used for the same purpose, but I don't remember anyone actually used it. You can see it tested in the image validation logic for devices, under the name "X of N".)

hughsie commented:
I've cleaned things up and reworked the API a little after trying to build a new plugin on top of it. I've switched around, so the plugin defines the children of a device, and just lets the daemon and client work out the single parent for each device. I think I'm happy with the patchset now.

hughsie commented on plugins/dell/fu-plugin-dell.c:
```diff
@@ -341,23 +340,22 @@ fu_plugin_dock_node (FuPlugin *plugin, GUsbDevice *device,
+					     component_desc);
+		fu_device_add_parent_guid (dev, DOCK_FLASH_GUID);
+	}
+	else
```
odd -- same line as brace and also braces for the lower half pls.

superm1 replied:
fixed

hughsie commented on plugins/synapticsmst/fu-plugin-synapticsmst.c:
```diff
@@ -140,6 +141,10 @@ fu_plugin_synaptics_add_device (FuPlugin *plugin,
 	fu_device_add_guid (dev, guid_str);
 
+	/* Currently recognizes TB16/WD15 */
+	if (data->dock_type)
```
`!= NULL` pls

superm1 replied:
fixed

superm1 commented:
So here's what it looks like with hooking up parents to the various stuff associated with Dell:

```
○
├─ XPS 13 9350 Thunderbolt Controller        016f9ebfa66e983c3933184e4c83a5c6b3f5b2b2
├─ XPS 13 9350 TPM 2.0                       eb4a4ec71c680a34c6248dc94ca017f81c772aa8
│ └─ XPS 13 9350 TPM 1.2                     b719feb08bda534ee0cb9df33b8de5c5458e2132
├─ XPS 13 9350 System Firmware               79f2996ff11608674c5d2e8db20e10e0b646f2cb
├─ Dell TB16                                 8e5d18ee7aa0401befa8a18c4d84c573ee249e22
│ ├─ Dell Thunderbolt Cable                  9d3474a5ba23256094c67a3ffbe6620312374c94
│ ├─ Dell Thunderbolt Dock                   b60958484239af7f9af424fa9144fa5fad241ddb
│ ├─ Dell TB16 Thunderbolt Cable             d4d8ddf1c3a265510eda5c2e00cf1fe621cc94e8
│ ├─ Dell TB16 Port Controller 1             a38cdf13b43897c26e698c5308073492bb32a449
│ ├─ Dell TB16 Port Controller 2             ecfba079d8c730bad07a099a9503ca1704a23a86
│ ├─ Synaptics VMM3330 inside Dell WD15/TB16 wired Dock 7b3cbc344094ace839263dd4711c84ea99bb6dc3
│ └─ Synaptics VMM3320 inside Dell WD15/TB16 wired Dock d83854e8c81f5d47b0b99c8a3b42ae5127c75eb8
└─ Iris Graphics 540                         8de6c7959053fd5798006dcc63590d33fa5e51cb
```

This is what it would be if you put a computer DMI title at top (of course not in right now).  If this was done then I think it makes sense to strip out the DMI names put on everything else.
```
○ Dell Inc. XPS 9350
├─ XPS 13 9350 Thunderbolt Controller        016f9ebfa66e983c3933184e4c83a5c6b3f5b2b2
├─ XPS 13 9350 TPM 2.0                       eb4a4ec71c680a34c6248dc94ca017f81c772aa8
│ └─ XPS 13 9350 TPM 1.2                     b719feb08bda534ee0cb9df33b8de5c5458e2132
├─ XPS 13 9350 System Firmware               79f2996ff11608674c5d2e8db20e10e0b646f2cb
├─ Dell TB16                                 8e5d18ee7aa0401befa8a18c4d84c573ee249e22
│ ├─ Dell Thunderbolt Cable                  9d3474a5ba23256094c67a3ffbe6620312374c94
│ ├─ Dell Thunderbolt Dock                   b60958484239af7f9af424fa9144fa5fad241ddb
│ ├─ Dell TB16 Thunderbolt Cable             d4d8ddf1c3a265510eda5c2e00cf1fe621cc94e8
│ ├─ Dell TB16 Port Controller 1             a38cdf13b43897c26e698c5308073492bb32a449
│ ├─ Dell TB16 Port Controller 2             ecfba079d8c730bad07a099a9503ca1704a23a86
│ ├─ Synaptics VMM3330 inside Dell WD15/TB16 wired Dock 7b3cbc344094ace839263dd4711c84ea99bb6dc3
│ └─ Synaptics VMM3320 inside Dell WD15/TB16 wired Dock d83854e8c81f5d47b0b99c8a3b42ae5127c75eb8
└─ Iris Graphics 540                         8de6c7959053fd5798006dcc63590d33fa5e51cb
```

hughsie commented:
> maybe adjusting DMI

I'll have a think about this some more next week. Thanks for all the fixes on the branch.
